### DISCUSSION TITLE
How does turbulence and gust enter the aircraft dynamics ?

### DISCUSSION THREAD

**Discussion by Dobid** [JSBSim-Team/jsbsim#1211] [Q&A] [answered]
Hello!
I'm going through a flight dynamics textbook (Small Unmanned Aircrafts by R.W. Beard) and I'm curious if the textbook describes how wind/turbulence/gust effects are effectively modeled in JSBSim.
Basically the textbook goes over on how the wind velocity is comprised of a steady ambient wind part and a stochastic part (generated with Dryden filters) which seems close to what JSBSim's FGWinds.cpp does. Then the textbook states that this wind velocity is used for computing the airspeed vector from which we can compute the airspeed norm $$V_a$$, the AoA $$\alpha$$ and the sideslip angle $$\beta$$. Which are used in the aerodynamic forces (lift, drag, side) and moments (l,m,n).
![image](https://github.com/user-attachments/assets/3baa9f98-2841-4e79-be9e-26626e952697)

I just want to make sure that in JSBSim, turbulence/gusts are only computed by the variables $$Va, \alpha, \beta$$ entering the aerodynamic forces and moments. Or if there's something else, maybe extra turbulence/gust forces and moments being directly applied to the aircraft's body.

Thanks in advance!

**Comment by Warning39** [comment]
Here is the relevant code which takes the wind into account:

https://github.com/JSBSim-Team/jsbsim/blob/318c80408b4566addf42983a764362edbddbb792/src/models/FGAuxiliary.cpp#L145-L167

**Reply by Dobid** [reply]
Ok so if I understand correctly, vAeroUVW is the velocity vector of the aircraft relative to the windspeed and vAeroPQR is the vector of angular rates including the turbulence angular rates.

vAeroUVW is used to compute $$V_a$$, $$\alpha$$ and $$\beta$$. Which are then used to compute the aerodynamic forces (lift, drag, side) and moments (rolling, pitching, yawing).

In addition, vAeroPQR angular rates, are also used in the computation of the aerodynamic forces and moments through p,q,r-aero-rad_sec properties.

In conclusion, the snippet of the book does not exactly describe how JSBSim manages wind effects on the aircraft. Since wind dependent turbulent angular velocities AeroPQR also enter the aerodynamic forces & moments calculations.

Hope I got everything right, please correct me if I made any mistakes as I'm still learning :)

**Comment by Warning39** [comment]
Yep, although your second reference to `vAeroUVW` should be `vAeroPQR`.

If you take a look at `FGWinds.cpp` you'll notice that `TurbPQR` as implied by the name is only set if some of the turbulence models are enabled.

Take a look at for example https://ntrs.nasa.gov/api/citations/19980028448/downloads/19980028448.pdf which mentions `PQR` in terms of the some of the turbulence models.

**Reply by Dobid** [reply]
Ok, thanks a lot !

**Comment by agodemar** [comment]
@Dobid 
My comment on AeroPQR, that is, the angular rates $(p_\mathrm{g}, q_\mathrm{g}, r_\mathrm{g})$ induced at the instantaneous aircraft CG location by a 'gust field'. 

In general, those rates should be time dependent, and more in general, they should be modified by the aircraft presence in the same field. 

As a first approximation, the gust field is considered 'frozen in time' therefore the rates $(p_\mathrm{g}, q_\mathrm{g}, r_\mathrm{g})$ can be pre-calculated and stored as functions of the position $(x_\mathrm{E}, y_\mathrm{E}, z_\mathrm{E})$ in the Earth-fixed frame, once the aircraft main dimensions are known (fuselage length, wing span, etc).

**Example of $q_\mathrm{g}$:** 

In the following schematics
![image](https://github.com/user-attachments/assets/160bce16-c056-4f80-9f19-beca017ee486)

$w_\mathrm{g}$ is a simple sinusoidal function of $x_\mathrm{E}$ (assuming that the CG is travelling towards the positive Earth-fixed $x$-axis). Along the longitudinal distance $\lambda_x$, the effect of such a law of $w_\mathrm{g}$ is, approximately, the same of __a positive aircraft rigid rotation, of pitch rate__ $q_\mathrm{g}$ __in a zero gust field__. In fact, the top part of the fuselage nose 'sees' a relative wind coming from above; the bottom part of the fuselage tail sees a relative wind coming from below. In this sense, someone calls the rate $q_\mathrm{g}$ an 'equivalent' rate.

Reference:
[Van Staveren, W. H. J. J., Analyses of Aircraft Responses to Atmospheric Turbulence, PhD Thesis, TUDelft, 2003](https://repository.tudelft.nl/record/uuid:c639ce70-4ea9-4c9b-8594-bb4b51e42c73)
See Chapter 5 for the theoretical explanations.
See Chapter 12, the "The Four Point Aircraft model" for a practical example.

**Comment by Warning39** [comment]
@Dobid you mentioned:

> and `vAeroPQR` is the vector of angular rates including the turbulence angular rates.

Rereading this and glancing at the JSBSim source code again, I'm not 100% certain what you meant. Just to be clear `vAeroPQR` is the angular rate of the aircraft's body relative to the air mass, the turbulence angular rates don't modify the aircraft's inertial angular rates.

See the following query on `AeroPQR` in the JSBSim repo to see what code makes use of `vAeroPQR`:

https://github.com/search?q=repo%3AJSBSim-Team%2Fjsbsim+AeroPQR&type=code

> In addition, `vAeroPQR` angular rates, are also used in the computation of the aerodynamic forces and moments through p,q,r-aero-rad_sec properties.

Although the `vAeroPQR` is exported via properties, this doesn't imply that they are used by JSBSim to compute aerodynamic forces or moments. In fact if you take a look at all the references to `AeroPQR` in the query link above you'll see that only the propellor code and rotor code make use of `vAeroPQR`. 

https://github.com/JSBSim-Team/jsbsim/blob/330a6e31ef26a8ce2427a9fa2279af14cd63322a/src/models/FGAuxiliary.cpp#L394-L396

**Reply by Dobid** [reply]
Sorry, my understanding of JSBSim's codebase might be limited.

> Although the vAeroPQR is exported via properties, this doesn't imply that they are used by JSBSim to compute aerodynamic forces or moments. In fact if you take a look at all the references to AeroPQR in the query link above you'll see that only the propellor code and rotor code make use of vAeroPQR.

When I look at Cessna 172p FDM .xml file I find this product for the "Lift due to pitch rate".
https://github.com/JSBSim-Team/jsbsim/blob/330a6e31ef26a8ce2427a9fa2279af14cd63322a/aircraft/c172p/c172p.xml#L704-L713
Taking the Lift force formula:
![image](https://github.com/user-attachments/assets/1b28d936-a46e-4bec-9ed7-ccd7475f5923)
I assume that "Lift due to pitch rate" is the following term: $$\frac{1}{2}\rho V_a^2 S C_{L_q} \frac{c}{2V_a} q$$ and in this case $$q$$ seems to be referenced with the property ``velocities/q-aero-rad_sec`` which directly comes from the vAeroPQR. So it makes me think that in fact, wind relative angular rates do enter aerodynamic forces/moments computation.

In my current experiments I'm using Milspec turbulence model so, I assume vAeroPQR is set here, in the case of Milspec turb:
https://github.com/JSBSim-Team/jsbsim/blob/330a6e31ef26a8ce2427a9fa2279af14cd63322a/src/models/atmosphere/FGWinds.cpp#L364-L369

Since:
https://github.com/JSBSim-Team/jsbsim/blob/318c80408b4566addf42983a764362edbddbb792/src/models/FGAuxiliary.cpp#L145-L147

and ``in.vUVW``, ``in.vPQR`` are described here:
https://github.com/JSBSim-Team/jsbsim/blob/330a6e31ef26a8ce2427a9fa2279af14cd63322a/src/models/FGAccelerations.h#L343-L346

I was assuming that ``vAeroUVW`` and `vAeroPQR` were the linear velocity and angular rates with respect to the wind.

**Comment by Warning39** [comment]
@Dobid talking of wind effects @agodemar was one of the authors on the following paper and implementation - [Flight Load Assessment for Light Aircraft Landing Trajectories in Windy Atmosphere and Near Wind Farms](https://www.mdpi.com/2226-4310/5/2/42).

In this case the wind vector field is calculated via CFD by OpenFOAM ahead of time, and then interpolated to generate a NED wind vector for each timestep for the aircraft's current position.

I've seen this sort of approach also applied to modelling the burble behind an aircraft carrier to see how it affects aircraft landing on the carrier.

**Comment by Warning39** [comment]
@Dobid there is a good description in the [CVN Airwake Modeling and Integration: Initial Steps in the Creation and Implementation of a Virtual Burble for F-18 Carrier Landing Simulations](https://sci-hub.se/10.2514/6.2005-6298) paper with regards to modelling direct moments from turbulence, similar to what @agodemar referred to in the thesis link above.

> The CFD airwake is integrated with the F-18
 aircraft simulation by applying instantaneous (time varying) three-component velocity data at the aircraft center of
 gravity (CG) and four other points at the extremities: nose, tail, and both wingtips. The velocities applied to the CG
 directly affect the u, v, and w body-axis velocity components of the aircraft, while the velocities applied to the
 extremities are used to calculate estimated rotational components.

> The structure of the F-18 aerodynamic model allows for only one angle of attack for all points on the airframe.
 With only u, v, and w components of airflow available, there will be some moments generated from the inherent
 aerodynamic and stability properties of the aircraft, but there can be no moments applied directly to the airframe
 unless some type of estimation is done. For example, suppose the aircraft CG passes directly through the center of a
 vortex that has an axis of rotation aligned with the aircraft longitudinal axis. Due to the upwash on one wing and
 downwash on the other, a rolling moment will result. If the simulation encounters the same situation, absent of any
 estimations, there would be zero effect on the airframe. Therefore, the three burble components were found at the
 “ends” of the aircraft longitudinal and lateral axes and divided by the perpendicular distance from the velocity vector
 and the CG of the aircraft.

As in the referenced F-18 aerodynamic model JSBSim also only calculates a single AoA and doesn't have any current support for creating additional direct moments from the turbulence models.

> In practice, the estimated rotational moments had little effect on the approach path. The F-18’s control surface power
 and excellent stability system minimized any glideslope or lineup excursions due to rotational airwake
 characteristics. The airwake velocities that exist at the CG were used to determine the total wind-relative velocity of
 the aircraft. In aircraft body axis, wind-relative velocities are used to calculate angle of attack and angle of sideslip
 for input into the aerodynamic model. The forces and moments calculated by the aerodynamic model are then
 passed to the equations of motion for integration.

**Comment by Warning39** [comment]
@Dobid I mentioned above:

> Although the `vAeroPQR` is exported via properties, this doesn't imply that they are used by JSBSim to compute aerodynamic forces or moments. In fact if you take a look at all the references to `AeroPQR` in the query link above you'll see that only the propellor code and rotor code make use of `vAeroPQR`. 

So it is the case that JSBSim doesn't internally replace the inertial rates `PQR` with `AeroPQR`, but generally aircraft modelers do make use of `AeroPQR`, e.g. from the 737 model in the repo:

https://github.com/JSBSim-Team/jsbsim/blob/330a6e31ef26a8ce2427a9fa2279af14cd63322a/aircraft/737/737.xml#L786-L796
 
https://github.com/JSBSim-Team/jsbsim/blob/330a6e31ef26a8ce2427a9fa2279af14cd63322a/aircraft/737/737.xml#L706-L716

https://github.com/JSBSim-Team/jsbsim/blob/330a6e31ef26a8ce2427a9fa2279af14cd63322a/aircraft/737/737.xml#L718-L728

**Reply by Dobid** [reply]
Yes that's what I remarked for the c172p in my last comment : https://github.com/JSBSim-Team/jsbsim/discussions/1211#discussioncomment-11699524
Maybe you'd prefer if I reply to your comments in a new comment and not as a reply in a thread.

**Comment by Warning39** [comment]
> Yes that's what I remarked for the c172p in my last comment

Yep, I saw that literally after I pushed `Comment` on my last post and saw you had made that comment 1min before, i.e. our comments passed each other on the internet 😉 

My general personal preference is to reply at the end of the comment stream as opposed to in a thread. But there is no hard and fast rule.

> So it makes me think that in fact, wind relative angular rates do enter aerodynamic forces/moments computation.

Yes, but only if the person developing the FDM for the particular aircraft type uses them. JSBSim doesn't internally automatically replace the inertial rates with the air relative versions for any of the aerodynamic force or moment calculations.

Unlike how turbulence was added to the F-18 HARV simulation - [Implementation and Testing of Turbulence Models for the F-18-HARV Simultion ](https://ntrs.nasa.gov/api/citations/19980028448/downloads/19980028448.pdf)

![image](https://github.com/user-attachments/assets/01286ee9-55a4-4497-b232-f9735bd8e01f)

**Comment by Dobid** [comment, marked as answer]
Ok so JSBSim does not impose the usage of `vAeroPQR` as an input of aerodynamic forces and moments.
It is up to the FDM designer to use PQR or AeroPQR in the aerodynamics calculations.
However, I've checked a couple of FDMs provided by JSBSim (F-16, C172P, Concorde...) and they all use AeroPQR, which also seems to be what the authors of the F-18 HARV simulation did.
I've also stumbled upon the [Matlab Dryden function](https://fr.mathworks.com/help/aeroblks/drydenwindturbulencemodeldiscrete.html) from the Aerospace toolbox. Which specifies turbulence angular rates as one of its outputs, just like what JSBSim does for the Milspec turbulence model.

So, is it safe to assume that there is a "consensus" for using wind relative PQR in the aerodynamic forces and moments as a way of simulating turbulence effects on an aircraft ? Because even if JSBSim doesn't explicitly make use of vAeroPQR in aerodynamic calculations, wouldn't it be odd to compute turbulence PQR (thus vAeroPQR) for not using it in the dynamics ?

**Comment by Warning39** [comment]
Yep, I'd say that's a fairly accurate summary. 

So whether it's a constant wind vector or a time/space varying wind vector it will change $V_a, \alpha, \beta$ which will result in a change in aerodynamic forces and moments. In addition, if any of the turbulence models are enabled then `AeroPQR` from the turbulence model will generate aerodynamic moments as long as the FDM designer has made use of the `AeroPQR` properties.

What JSBSim doesn't model is additional direct aerodynamic moments due to varying wind velocities along the wingspan and length of the aircraft like some simulations, e.g. the F-18 carrier landing paper I mentioned above.
